### DISCUSSION TITLE
Welcome to poster-design Discussions!

### DISCUSSION THREAD

**Discussion by palxiao** [palxiao/poster-design#22] [Announcements]
你好！这里是一个开放的讨论区域，你可以提出问题或分享想法，与其他社区成员互动。一些项目的重要公告也可能在此发布。

期待与你一起共建💪

## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by guojiawei001** [comment]
使用这个在线截图工具生成的图片如果背景底色是透明的就会有问题

**Reply by palxiao** [reply]
细啊，我已修复这个问题，并且现在支持透明背景png下载

**Reply by guojiawei001** [reply]
你好，能请教下您修改的地方是哪块吗？我虽然调整的背景色变成透明了，（之前如果要调成透明的话底部背景色就变成黑色了），但是在生成图片的时候并不是透明的，就很奇怪

**Reply by palxiao** [reply]
https://github.com/palxiao/poster-design/discussions/25

**Reply by guojiawei001** [reply]
另外也建议添加一个一键清空画布的功能，想问下这些在线资源的图片是您那边在不断的更新是吗？我看图片上一些内容中svg啊图片啊这些元素资源池中并没有，不知道您这边是怎么传上去的呢

**Reply by palxiao** [reply]
这个功能实际新建一个空白画布即可。在线Demo中的模板和素材资源仅为演示作用，来源于网络收集，我是直接sql入库的。正常上传模板和素材资源应该由设计师完成，例如在psd解析中调整并上传模板，只是我没做资源管理，所以仅保存在用户作品中，但业务逻辑是类似的。

**Reply by guojiawei001** [reply]
哦不过网上资源一般都是png或者jpeg的格式的吧，然后您找到这种图片进行二次切割然后入库的吗？还是说您直接网上资源找到就是这种json格式写好的，直接入的数据库呢？

**Reply by guojiawei001** [reply]
另外大神您别嫌我烦哈，再请教你个问题，我这边使用vm架设了个虚拟主机，然后配置的centos7也安装了node和npm这些，我进入到\screenshot这个目录后通过npm run build将程序打包好，然后进入到这个dist目录中，通过node server.js是可以运行的，但是只要一运行保存图片的操作就会直接报：/www/wwwroot/www.zxzt.com/scimg/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:197
            reject(new Error([
                   ^

Error: Failed to launch the browser process! spawn /opt/google/chrome-unstable/chrome ENOENT


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md

    at onClose (/www/wwwroot/www.zxzt.com/scimg/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:197:20)
    at ChildProcess.<anonymous> (/www/wwwroot/www.zxzt.com/scimg/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:189:85)这个错误，
而如果我进入到上级目录执行：npm run dev那么它就可以通过我的接口下载成功这个图片，不知道什么原因

**Reply by palxiao** [reply]
网络收集只是一种隐晦的说法，直白点就是爬虫，数据清洗成自己的之后入库。

**Reply by palxiao** [reply]
你不能在 dist 目录中直接运行 pm2，因为webpack打包时已经配置依赖externals了，应该将打包好的 server.js 移动到上级目录，也就是和 node_modules 同级，这点我在 readme 中其实有详细说明的

**Reply by guojiawei001** [reply]
谢谢大佬的耐心解答我在服务器上直接把node_modules删除，然后运行npm install,接着执行npm run build,最后将dist里的server.js复制到和node_modules下面，然后执行node server.js这个流程是否有问题呢？我这样操作了还是一样的报错就比较奇怪了，我的是centos7的服务器，我又重新安装了下chrome,路径是：/usr/bin/google-chrome然后我启动这个保存图片后直接提示：
Error: Failed to launch the browser process!
mkdir: 无法创建目录"/root": 权限不够
touch: 无法创建"/root/.local/share/applications/mimeapps.list": 权限不够
[0919/142826.641720:ERROR:headless_shell.cc(250)] Multiple targets are not supported.


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md

    at onClose (/www/wwwroot/www.zxzt.com/scimg_new/node_modules/puppeteer/src/node/BrowserRunner.ts:223:9)
    at ChildProcess.<anonymous> (/www/wwwroot/www.zxzt.com/scimg_new/node_modules/puppeteer/src/node/BrowserRunner.ts:210:61)
    at ChildProcess.emit (node:events:406:35)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
[ERROR] 14:28:26 Error: Failed to launch the browser process!
mkdir: 无法创建目录"/root": 权限不够
touch: 无法创建"/root/.local/share/applications/mimeapps.list": 权限不够
[0919/142826.641720:ERROR:headless_shell.cc(250)] Multiple targets are not supported.


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md

**Reply by palxiao** [reply]
权限不够加个sudo试试呢？还有通常chrome默认的安装路径为：/opt/google/chrome-unstable/chrome，你去opt目录下看看有没有，一般不会是装在usr里的，那可能只是个软链。

**Reply by guojiawei001** [reply]
[root@localhost scimg_new]# cd /opt/google/chrome-unstable
-bash: cd: /opt/google/chrome-unstable: 没有那个文件或目录
[root@localhost scimg_new]# 我直接进这个目录都找不到这个目录的，
[root@localhost scimg_new]# which google-chrome
/usr/bin/google-chrome这里看到的目录是这样的，可能centos和您的服务器不一样并把，算了先不管这个了，晚点我自己在研究下看看，就让他npm run dev跑着吧先，这样跑起来之后我其他地方调用这个来生成图片就没问题了，太感谢大神的耐心解答了，之后使用过程中有什么问题我在反馈给您哈，希望别嫌我烦

**Reply by palxiao** [reply]
也有可能是虚拟机的问题，之前我看别人也有在centos服务器上成功跑起项目的，这个应该不是大问题。

**Reply by guojiawei001** [reply]
我执行了下sudo yum install google-chrome-stable
直接npm run dev都报错了，然后我就把服务器镜像混滚了一下，我服务器上并没有安装pm2的node项目管理工具,也就是说我无法执行：pm2 start server.js这个命令而是使用的node server.js我估计跟这个有关系吧，我服务器上配置的是nginx的环境

**Reply by palxiao** [reply]
这个应该没关系的，pm2的作用只是守护进程，其它跟node执行是一样的

**Reply by guojiawei001** [reply]
那确实奇怪，就现在我的这个centos系统里依旧没有'/opt/google/chrome-unstable/chrome',这个文件，但是执行npm run dev却奇迹般的可以保存图片并下载，

**Reply by guojiawei001** [reply]
大神,比较奇怪我看你更新的里面说的是修改：src/assets/styles/design.less，src/assets/styles/main.less，src/views/Index.vue这三个文件就可以么是吧？我直接下载了新版然后把这三个文件替换了一下，可是为啥生成的图片背景色依旧不是透明，就很奇怪呢，我看
exports.drawLink = 'https://design.palxp.cn/draw'这一步影响的，如果我使用这个代码的话就不会有问题，生成就是透明的，但是我如果把接口改成我自己的就不是透明的了

**Reply by palxiao** [reply]
screenshot也有更改的，不止改的前端文件，你再仔细看下commit

**Reply by guojiawei001** [reply]
我是把最新的screenshot这个目录单独拿出来放到服务器上然后直接执行的npm install 和npm run dev然后本地的程序直接调取的：http://虚拟机内网的id:7001的SCREEN_URL:'http://192.168.3.236:7001'这样的

**Reply by guojiawei001** [reply]
因为我本地是windows的这个screenshot跑不起来，然后我就用VMware搭了一个虚拟的centos虚拟机

**Reply by palxiao** [reply]
我还是怀疑虚拟机的问题，你换成我在线demo的url截图没问题那么代码本身就没有问题

**Reply by guojiawei001** [reply]
嗯screenshot这个代码中有一段是最后生成图片的接口：https://design.palxp.cn/draw使用你的这个接口就是透明的，不过我在部署一份你的最新的这套程序的话，然后调用https://192.168.3.233/draw也就是我的本地windows环境的话就没有问题也是透明的了（我的虚拟机和windows之间可以相互访问的）我估计是我的代码可能还有地方需要修改吧，我在排查排查看看吧

**Reply by guojiawei001** [reply]
找到打包生成时报错的原因了把uitls/download-single.ts中的executablePath: isDev ? null : executablePath,直接改成executablePath: null,就可以了，他这个executablePath应该是内部自己会自动获取的

**Reply by guojiawei001** [reply]
那个最终图片背景不透明的问题我也找到了，是因为之前的文件里面的app.vue中的style有问题
<style lang="less">
#appindex {
  min-width: 1180px;
  .viewWrap {
    background-color: #ffffff;
    min-height: calc(110vh - 110px);
    min-width: 1170px;
  }
}把这个 background-color: #ffffff;删掉就好了

**Reply by guojiawei001** [reply]
建议大神往那个修复方法中加一下

**Reply by palxiao** [reply]
都有加的哈，提交了两个相关的commit。稳定版本发布前建议直接拉取主分支最新代码规避这些问题。

**Reply by guojiawei001** [reply]
哦那个更新里面有写，大神我问下，如果要生成的图片大概是3779*9448这样的大尺寸图片，好像生成出来的尺寸就会出问题，这个是为什么呢？我看程序里也没有对尺寸这块的限制的呀

**Reply by palxiao** [reply]
长图合成会宕机，代码里有限制，你没拉最新的代码

**Reply by guojiawei001** [reply]
哦是了好像，const maxPXs = 42118400 // 超出此规格会触发限制器降低dpr，节省服务器资源
const maximum = 5000 把这两个值搞大点就好了，试了下我这个尺寸还可以不怎么会宕机，估计生成的多了可能会吧

**Reply by guojiawei001** [reply]
对了大神这个如果保存自己做好的图片好像不会生成缩略图吧，在我的作品中只能展示出作品名称，而没有缩略图吗？

**Reply by palxiao** [reply]
为了节省资源我只对模板做了缩略图，作品里的没有，在保存时后台调一下生成图片的接口然后存url到作品的cover字段就行。https://xp.palxp.cn/apidoc/screenshot.html#api-screenShot-GetApiScreenshots

**Reply by guojiawei001** [reply]
大佬再咨询你个问题，为什么我修改了代码想让第一次打开的时候先默认隐藏推荐模板的展示，然后点击一下推荐模板在显示对应的信息，
我传递了一个
 props: {  
  setactive: {  
      type: Boolean,  
      default: true,  
    },  
  }, 然后在const state = reactive({
      widgetClassifyList: widgetClassifyListData,
      activeWidgetClassify: -1,
      active:props.setactive,
    })这里设置时默认先把这个active隐藏掉，可是为什么点击一下模板却不展示内容呢？必须点几下收起的那个小图标再打开下才会显示

**Reply by palxiao** [reply]
不太清楚你的需求，但是根据你描述的情况，我觉得在onMounted钩子中设置active为false就好了，state中的active默认值保持为true

**Comment by guojiawei001** [comment]
我是这样设置的，但是如果这个action默认为false的情况下点下那个模板的字样，展开就是空白的

**Reply by guojiawei001** [reply]
然后再点下那个关闭的小箭头，然后再点那个模板就显示又信息了

**Reply by guojiawei001** [reply]
大佬如果我在外部调用的素材接口，我想再列表中给每个素材加一个下载的功能，如何在不走编辑器的情况下直接下载？有对应的接口吗？

**Reply by palxiao** [reply]
没有素材管理相关的功能，从外部下载的话直接下载素材的封面图就行了吧

**Reply by guojiawei001** [reply]
哦这个素材封面好像是缩略图是不是？直接走/api/screenshots?id=${id}&width=${width}&height=${height}这个地址是不是就行？不过这个id好像对应的不是templateid

**Reply by palxiao** [reply]
素材也是有宽高这些属性的，逻辑是差不多的，只是我没写这部分代码。我的意思是你要从外部下载的话，资源内容是不会变化的吧，就不用每次都去调用screenshots产生额外的开销，暴露出生成后的url就行了

**Reply by guojiawei001** [reply]
大佬这个生成后的url连接是什么呢？比如模板id是1120的这个模板素材，那个cover?

**Reply by palxiao** [reply]
生成后资源保存在哪里url就是什么呀，目前项目中没有素材管理的功能，所以不能直接生成素材封面的

**Reply by guojiawei001** [reply]
你的意思是static或者cache目录下面的那些生成后的素材文件？如果想下载你展示的这些模板的话我是不是需要重新都下载一遍才能生成独赢的url呢？另外https://app.palxp.com:7002/api/screenshots?id=1125&width=1242&height=2208&r=0.2056305948725654这个接口中传入的id不是模板的id呀？我想点击下下载如果static目录下没有这个文件然后走下这个接口来生成是否可行呢？

**Reply by palxiao** [reply]
目前的接口仅支持作品下载，传入的是作品id或模板id，没有渲染素材id的逻辑，这是要额外开发的

**Reply by guojiawei001** [reply]
好的，我这边看下吧，貌似走下save之后生成出来的那个id才是正确的，通过那个id然后渲染下走下screenshots这个接口生成下图片应该就可以了
